File: with_lib/lll_custom_shell.c
#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <time.h>
#include <limits.h>
#include <inttypes.h>
#include <stdint.h>

#ifdef PATH_MAX
#ifdef NAME_MAX
#define FILE_SYSTEM_BUFF_SIZE (PATH_MAX + NAME_MAX + 1)
#else
#error "NAME_MAX not defined"
#endif
#else
#error "PATH_MAX not defined"
#endif

#define NR_OF_SUPPORTET_LANG 11
#define CARD_LEN 15
#define MAX_SIZE_ELEMENT 15
#define SUPPORTET_GROUPS 11
#define ERRR 0xFFFAFFFu

typedef struct {
    int doppel;
    const char *name;
    uint64_t countet_size;
    int globalsumindex;
} Type;

typedef struct MemNode {
    void *ptr;
    struct MemNode *next;
} MemNode;

static int unix_index = 0;
static int unix_max = 1024;
static MemNode *mem_head = NULL;

static Type types[NR_OF_SUPPORTET_LANG] = {
    {0, "txt",    0ULL, 0},   // 0: PHP
    {0, "c",      0ULL, 1},   // 1: C
    {0, "cpp",    0ULL, 2},   // 2: C++
    {0, "ino",    0ULL, 3},   // 3: ESP
    {0, "py",     0ULL, 4},   // 4: Python
    {0, "java",   0ULL, 5},   // 5: Java
    {0, "sh",     0ULL, 6},   // 6: Shell
    {0, "js",     0ULL, 7},   // 7: JavaScript
    {0, "html",   0ULL, 8},   // 8: HTML
    {0, "asm",    0ULL, 9},   // 9: Assembly
    {0, "rs",     0ULL, 10}   // 10: Rust
};

static const char *nameList[NR_OF_SUPPORTET_LANG] = {
    "undefined",    // 0
    "C",            // 1
    "C++",          // 2
    "ESP",          // 3
    "Python",       // 4
    "Java",         // 5
    "ShellScript",  // 6
    "JavaScript",   // 7
    "HTML",         // 8
    "Assembly",     // 9
    "Rust"          // 10
};

static const uint64_t colorList[NR_OF_SUPPORTET_LANG] = {
    0x2222077877CC,   // 0: PHP
    0x222205666666,   // 1: C
    0x22220A2222FF,   // 2: C++
    0x22220FEEAA00,   // 3: ESP
    0x222200A8A862,   // 4: Python
    0x222200FFFF00,   // 5: Java
    0x222200FF0010,   // 6: ShellScript
    0x222200BB0010,   // 7: JavaScript
    0x222200888888,   // 8: HTML
    0x222200FF8800,   // 9: Assembly
    0x222200DEA584    // 10: Rust
};

void clipper(char *content) {
    size_t len = strlen(content);
    FILE *p = popen("xclip -selection clipboard", "w");
    int fd = fileno(p);
    ssize_t w = write(fd, content, len);
    pclose(p);
}

/* Convert mode bits to string like ls -l */
static void mode_to_str(mode_t m, char out[11]) {
    out[0] = S_ISDIR(m) ? 'd' : S_ISLNK(m) ? 'l' : S_ISCHR(m) ? 'c' :
             S_ISBLK(m) ? 'b' : S_ISFIFO(m) ? 'p' : S_ISSOCK(m) ? 's' : '-';
    out[1] = (m & S_IRUSR) ? 'r' : '-';
    out[2] = (m & S_IWUSR) ? 'w' : '-';
    out[3] = (m & S_IXUSR) ? 'x' : '-';
    out[4] = (m & S_IRGRP) ? 'r' : '-';
    out[5] = (m & S_IWGRP) ? 'w' : '-';
    out[6] = (m & S_IXGRP) ? 'x' : '-';
    out[7] = (m & S_IROTH) ? 'r' : '-';
    out[8] = (m & S_IWOTH) ? 'w' : '-';
    out[9] = (m & S_IXOTH) ? 'x' : '-';
    out[10] = '\0';
}

static int isDoubleIndex(int ind) {
    return types[ind].doppel;
}

static void process_size(off_t size, char *buf, size_t buflen) {
    const char *units[] = {"B","KB","MB","GB","TB"};
    double s = (double)size;
    int u = 0;
    while (s >= 1024 && u < 4) {
        s /= 1024;
        u++;
    }
    snprintf(buf, buflen, "%.1f %s", s, units[u]);
}

int memlist_add(void *ptr) {
    if (!ptr) return -1;
    MemNode *n = malloc(sizeof(MemNode));
    if (!n) return -1;
    n->ptr = ptr;
    n->next = mem_head;
    mem_head = n;
    return 0;
}

void memlist_free_all() {
    MemNode *cur = mem_head;
    while (cur) {
        MemNode *next = cur->next;
        if (cur->ptr) free(cur->ptr);
        free(cur);
        cur = next;
    }
    mem_head = NULL;
}

char *coloramaprintifarma(const char *ptext, uint64_t col) {
    uint8_t bg_r = (col >> 40) & 0xFF;
    uint8_t bg_g = (col >> 32) & 0xFF;
    uint8_t bg_b = (col >> 24) & 0xFF;
    uint8_t fg_r = (col >> 16) & 0xFF;
    uint8_t fg_g = (col >> 8)  & 0xFF;
    uint8_t fg_b = col & 0xFF;
    const char *suffix = "\033[0m";
    const char *fmt = "\033[38;2;%u;%u;%um\033[48;2;%u;%u;%um";
    int prefix_len = snprintf(NULL, 0, fmt,
                              fg_r, fg_g, fg_b,
                              bg_r, bg_g, bg_b);
    if (prefix_len < 0) return NULL;
    size_t text_len   = strlen(ptext);
    size_t suffix_len = strlen(suffix);
    size_t total_len = (size_t)prefix_len + text_len + suffix_len + 1;
    char *buf = malloc(total_len);
    if (!buf) return NULL;
    memlist_add(buf);
    int written = snprintf(buf, total_len, fmt,
                           fg_r, fg_g, fg_b,
                           bg_r, bg_g, bg_b);
    memcpy(buf + written, ptext, text_len);
    memcpy(buf + written + text_len, suffix, suffix_len + 1);
    return buf;
}

static size_t utf8_char_len(unsigned char c) {
    if (c < 0x80) return 1;
    if ((c >> 5) == 0x6) return 2;
    if ((c >> 4) == 0xE) return 3;
    if ((c >> 3) == 0x1E) return 4;
    return 1;
}

static char* coloramaprintifarmafillup(const char *text, uint64_t advansed_color) {
    if (!text) return NULL;
    char *res = malloc(CARD_LEN + 1);
    if (!res) return NULL;
    memlist_add(res);
    size_t pos = 0;
    size_t i = 0;
    while (pos < CARD_LEN && text[i]) {
        size_t clen = utf8_char_len((unsigned char)text[i]);
        if (pos + clen > CARD_LEN) break;
        memcpy(res + pos, text + i, clen);
        pos += clen;
        i += clen;
    }
    while (pos < CARD_LEN) {
        res[pos++] = ' ';
    }
    res[CARD_LEN] = '\0';
    char *colored = coloramaprintifarma(res, advansed_color);
    return colored;
}

static char *coll(uint64_t col, const char *ptext) {
    uint8_t r = (col >> 16) & 0xFF;
    uint8_t g = (col >> 8)  & 0xFF;
    uint8_t b = col         & 0xFF;
    const char *suffix = "\033[0m";
    const char *fmt = "\033[38;2;%u;%u;%um";
    int prefix_len = snprintf(NULL, 0, fmt, r, g, b);
    if (prefix_len < 0) return NULL;
    size_t text_len   = strlen(ptext);
    size_t suffix_len = strlen(suffix);
    size_t total_len = (size_t)prefix_len + text_len + suffix_len + 1;
    char *buffer = malloc(total_len);
    if (!buffer) return NULL;
    memlist_add(buffer);
    int written = snprintf(buffer, total_len, fmt, r, g, b);
    memcpy(buffer + written, ptext, text_len);
    memcpy(buffer + written + text_len, suffix, suffix_len + 1);
    return buffer;
}

static char *chngg(const char *pp, uint64_t color, int more_space) {
    int mse = (more_space == 1) ? (MAX_SIZE_ELEMENT * 3) : MAX_SIZE_ELEMENT;
    char *rptr = malloc(mse + 1);
    if (!rptr) return NULL;
    memlist_add(rptr);
    size_t pos = 0;
    size_t i = 0;
    while (pos < mse && pp[i]) {
        size_t clen = utf8_char_len((unsigned char)pp[i]);
        if (pos + clen > mse) break;
        memcpy(rptr + pos, pp + i, clen);
        pos += clen;
        i += clen;
    }
    while (pos < mse) {
        rptr[pos++] = ' ';
    }
    rptr[mse] = '\0';
    char *rrr = coll(color, rptr);
    return rrr;
}

static char *chngg_wrapper(int index, uint64_t size) {
    return coloramaprintifarmafillup(nameList[index], colorList[index]);
}

static int get_type_ind_by_name(char *type_s) {
    for (int i = 1; i < NR_OF_SUPPORTET_LANG; ++i) {
        if (strcmp(types[i].name, type_s) == 0) {
                return i;
            }
    }
    return ERRR;
}

static int add_type_value(int pp_index, size_t filesize) {
    uint64_t addid = (uint64_t)filesize;
    types[pp_index].countet_size += addid;
    return 0;
}

static void get_type(const char *fullpath, int size, size_t filesize) {
    int point_index;
    int flag = 0;
    for (point_index = size - 1; point_index >= 0; --point_index) {
        if (fullpath[point_index] == '.') { flag = 1; break; }
        if (fullpath[point_index] == '/') { flag = 0; break; }
    }
    if (!flag) return;
    int len = size - point_index - 1;
    if (len <= 0) return;
    char *type_s = malloc(len + 1);
    if (!type_s) return;
    for (int i = 0; i < len; ++i) type_s[i] = fullpath[point_index + 1 + i];
    type_s[len] = '\0';
    unsigned int index = get_type_ind_by_name(type_s);
    if (index != ERRR) add_type_value(index, filesize);
    free(type_s);
}

static off_t folder_size(const char *path) {
    DIR *d = opendir(path);
    if (!d) {
        perror("opendir");
        return 0;
    }
    off_t total = 0;
    struct dirent *e;
    while ((e = readdir(d)) != NULL) {
        if (strcmp(e->d_name, ".") == 0 || strcmp(e->d_name, "..") == 0) continue;
        char fullpath[PATH_MAX];
        if (snprintf(fullpath, sizeof(fullpath), "%s/%s", path, e->d_name) >= (int)sizeof(fullpath)) continue;
        struct stat st;
        if (lstat(fullpath, &st) == -1) {
            perror("lstat");
            continue;
        }
        if (S_ISDIR(st.st_mode)) {
            total += folder_size(fullpath);
        } else {
            total += st.st_size;
            get_type(fullpath, (int)strlen(fullpath), st.st_size);
        }
    }
    closedir(d);
    return total;
}

static void getDomLangIndex(int *index, uintptr_t *size) {
    uint64_t local_sums[SUPPORTET_GROUPS] = {0};
    for (int i = 1; i < NR_OF_SUPPORTET_LANG; ++i) {
        int group = types[i].globalsumindex;
        if (group >= 0 && group < SUPPORTET_GROUPS) {
            local_sums[group] += types[i].countet_size;
        }
    }
    int maxIndex = 0;
    uint64_t maxSize = 0;
    for (int i = 0; i < SUPPORTET_GROUPS; ++i) {
        if (local_sums[i] > maxSize) { maxSize = local_sums[i]; maxIndex = i; }
    }
    *index = maxIndex;
    *size = maxSize;
    for (int i = 0; i < NR_OF_SUPPORTET_LANG; ++i) types[i].countet_size = 0;
}

void getAbsolutPath(char *name, int nameLen) {
    char path[FILE_SYSTEM_BUFF_SIZE];
    if (getcwd(path, sizeof(path)) != NULL) {
        int pathLen = strlen(path);
	int len = pathLen + nameLen + 2;
	char *absolutPath = malloc(len + 1);
	absolutPath[len] = '\0';
	int k = 0;
	for (int i = 0; i < len; ++i) {
	    if (i < pathLen) {
	        absolutPath[i] = path[k];
		++k;
	    }
	    if (i > pathLen) {
	        absolutPath[i] = name[k];
		++k;
	    }
	    if (i == pathLen) {
	        absolutPath[i] = '/';
		k = 0;
	    }
	}
	printf("%s\n", absolutPath);
	clipper(absolutPath);
	free(absolutPath);
    } else {
        perror("ERROR: getcwd");
    }
}

void addUnixTimestamp(time_t unixTimestamp, time_t **unixTimestamps) {
    if (*unixTimestamps == NULL) {
        *unixTimestamps = malloc(unix_max * sizeof(time_t));
        if (*unixTimestamps == NULL) {
            perror("malloc unixTimestamps");
            return;
        }
    }
    (*unixTimestamps)[unix_index] = unixTimestamp;
    unix_index++;
    if (unix_index == unix_max) {
        time_t *copy = malloc(unix_max * sizeof(time_t));
        if (copy == NULL) {
            perror("malloc copy");
            return;
        }
        for (int i = 0; i < unix_max; i++) {
            copy[i] = (*unixTimestamps)[i];
        }
        free(*unixTimestamps);
        unix_max *= 2;
        *unixTimestamps = malloc(unix_max * sizeof(time_t));
        if (*unixTimestamps == NULL) {
            perror("malloc new unixTimestamps");
            free(copy);
            return;
        }
        for (int i = 0; i < unix_max / 2; i++) {
            (*unixTimestamps)[i] = copy[i];
        }
        free(copy);
    }
}

typedef struct {
    time_t unixtimestamp;
    char *row;
} TTAS;

void swap(TTAS *a, TTAS *b) {
    TTAS temp = *a;
    *a = *b;
    *b = temp;
}

void reverseArray(TTAS *arr, int n) {
    int left = 0;
    int right = n - 1;
    while (left < right) {
        swap(&arr[left], &arr[right]);
        left++;
        right--;
    }
}

/* 1. Sortieren: neueste zuerst (wie ls -t) */
void sortByTimestamp(time_t *unixTimestamps, char **rows, int n) {
    if (unixTimestamps == NULL || rows == NULL || n <= 1)
        return;
    for (int i = 0; i < n - 1; ++i) {
        int maxIndex = i;
        for (int j = i + 1; j < n; ++j) {
            if (unixTimestamps[j] > unixTimestamps[maxIndex]) {
                maxIndex = j;
            }
        }
        if (maxIndex != i) {
            time_t tempTime = unixTimestamps[i];
            unixTimestamps[i] = unixTimestamps[maxIndex];
            unixTimestamps[maxIndex] = tempTime;
            char *tempRow = rows[i];
            rows[i] = rows[maxIndex];
            rows[maxIndex] = tempRow;
        }
    }
    TTAS *array = malloc(n * sizeof(TTAS));
    if (array == NULL) {
        perror("malloc TTAS array");
        return;
    }
    for (int i = 0; i < n; ++i) {
        array[i].unixtimestamp = unixTimestamps[i];
        array[i].row           = rows[i]; // copy *ptr
    }
    reverseArray(array, n);
    for (int i = 0; i < n; ++i) {
        unixTimestamps[i] = array[i].unixtimestamp;
        rows[i]           = array[i].row;
    }
    free(array);
}

int main_lll() {
    DIR *d = opendir(".");
    if (d == NULL) {
        perror("opendir");
        return 1;
    }
    struct dirent *e;
    char **rows = NULL;
    time_t *unixTimestamps = NULL;
    size_t count = 0, cap = 0;
    while ((e = readdir(d)) != NULL) {
        if (strcmp(e->d_name, ".") == 0 || strcmp(e->d_name, "..") == 0)
            continue;
        char path[PATH_MAX];
        if (snprintf(path, sizeof(path), "./%s", e->d_name) >= (int)sizeof(path))
            continue;
        struct stat st;
        if (lstat(path, &st) == -1) {
            perror("lstat");
            continue;
        }
        char perms[11];
        mode_to_str(st.st_mode, perms);
        nlink_t links = st.st_nlink;
        struct passwd *pw = getpwuid(st.st_uid);
        struct group  *gr = getgrgid(st.st_gid);
        const char *owner = pw ? pw->pw_name : "unknown";
        const char *group = gr ? gr->gr_name : "unknown";
        char sizebuf[32];
        char timestr[64];
        struct tm tm;
        localtime_r(&st.st_mtime, &tm);
        strftime(timestr, sizeof(timestr), "%e. %b %H:%M", &tm);
        time_t timestamp = st.st_mtime;
        char dtype = S_ISDIR(st.st_mode) ? 'f' : 'd';
        if (S_ISDIR(st.st_mode)) {
            off_t sz = folder_size(path);
            process_size(sz, sizebuf, sizeof(sizebuf));
            get_type(path, (int)strlen(path), (size_t)sz);
        } else {
            process_size(st.st_size, sizebuf, sizeof(sizebuf));
            get_type(path, (int)strlen(path), (size_t)st.st_size);
        }
        int ja_index = 0;
        uintptr_t ja_size = 0;
        getDomLangIndex(&ja_index, &ja_size);
        char line[FILE_SYSTEM_BUFF_SIZE];
        if (S_ISLNK(st.st_mode)) {
            char target[PATH_MAX];
            ssize_t len = readlink(path, target, sizeof(target) - 1);
            if (len != -1) {
                target[len] = '\0';
                snprintf(line, sizeof(line), "%s %2lu %s %s %8s %s %s -> %s",
                         perms, (unsigned long)links, owner, group, sizebuf,
                         timestr, e->d_name, target);
            } else {
                snprintf(line, sizeof(line), "%s %2lu %s %s %8s %s %s",
                         perms, (unsigned long)links, owner, group, sizebuf,
                         timestr, e->d_name);
            }
        } else {
            snprintf(line, sizeof(line), "%s %2lu %s %s %8s %s %s %s",
                     perms, (unsigned long)links,
                     chngg(owner,     0x44FF44, 0),
                     chngg(group,     0x88FF88, 0),
                     chngg(sizebuf,   0xFF0000, 0),
                     chngg(timestr,   0x888866, 0),
                     chngg_wrapper(ja_index, ja_size),
                     chngg(e->d_name, (dtype == 'f') ? 0x9999FF : 0xFF33FF, 1));
        }
        if (count + 1 > cap) {
            size_t ncap = cap ? cap * 2 : 64;
            char **tmp = realloc(rows, ncap * sizeof(char*));
            if (!tmp) {
                perror("realloc rows");
                break;
            }
            rows = tmp;
            cap = ncap;
        }
        rows[count] = strdup(line);
        if (!rows[count]) {
            perror("strdup");
            break;
        }
        addUnixTimestamp(timestamp, &unixTimestamps);
        count++;
    }
    closedir(d);
    sortByTimestamp(unixTimestamps, rows, (int)count);
    for (size_t i = 0; i < count; ++i) {
        puts(rows[i]);
        free(rows[i]);
    }
    free(rows);
    free(unixTimestamps);
    return 0;
}

int main(int argc, char **argv) {
    if (argc == 1) {
        main_lll();
	memlist_free_all();
    }
    if (argc == 2) {
	if (memcmp(argv[1], "max", 3) == 0) {
	    printf("MAX PATH LENGTH: %ld\n", FILE_SYSTEM_BUFF_SIZE);
	    return 0;
	}
        getAbsolutPath(argv[1], strlen(argv[1]));
    }
    return 0;
}

File: with_lib/lll_custom_shell.rs
use std::env;
use std::fs;
use std::os::unix::fs::{MetadataExt, PermissionsExt};
use std::io::Write;
use std::process::{Command, Stdio};
use std::path::Path;

const NR_OF_SUPPORTED_LANG: usize = 11;
const CARD_LEN: usize = 15;
const MAX_SIZE_ELEMENT: usize = 15;
const SUPPORTED_GROUPS: usize = 11;
const ERRR: u32 = 0xFFFAFFF;

#[derive(Clone, Copy)]
struct TypeInfo {
    name: &'static str,
    counted_size: u64,
    globalsumindex: usize,
}

static mut TYPES: [TypeInfo; NR_OF_SUPPORTED_LANG] = [
    TypeInfo { name: "",    counted_size: 0, globalsumindex: 0 },
    TypeInfo { name: "c",   counted_size: 0, globalsumindex: 1 },
    TypeInfo { name: "cpp", counted_size: 0, globalsumindex: 2 },
    TypeInfo { name: "ino", counted_size: 0, globalsumindex: 3 },
    TypeInfo { name: "py",  counted_size: 0, globalsumindex: 4 },
    TypeInfo { name: "java",counted_size: 0, globalsumindex: 5 },
    TypeInfo { name: "sh",  counted_size: 0, globalsumindex: 6 },
    TypeInfo { name: "js",  counted_size: 0, globalsumindex: 7 },
    TypeInfo { name: "html",counted_size: 0, globalsumindex: 8 },
    TypeInfo { name: "asm", counted_size: 0, globalsumindex: 9 },
    TypeInfo { name: "rs",  counted_size: 0, globalsumindex: 10 },
];

static NAME_LIST: [&str; NR_OF_SUPPORTED_LANG] = [
    "undefind", "C", "C++", "ESP", "Python", "Java", "ShellScript",
    "JavaScript", "HTML", "Assembly", "Rust",
];

static COLOR_LIST: [u64; NR_OF_SUPPORTED_LANG] = [
    0x2222077877CC, 0x222205666666, 0x22220A2222FF, 0x22220FEEAA00,
    0x222200A8A862, 0x222200FFFF00, 0x222200FF0010, 0x222200BB0010,
    0x222200888888, 0x222200FF8800, 0x222200DEA584,
];

fn mode_to_str(mode: u32) -> String {
    let mut s = String::with_capacity(11);
    s.push(match mode & 0o170000 {
        0o040000 => 'd', 0o120000 => 'l', 0o020000 => 'c',
        0o060000 => 'b', 0o010000 => 'p', 0o140000 => 's',
        _ => '-',
    });
    s.push(if mode & 0o400 != 0 { 'r' } else { '-' });
    s.push(if mode & 0o200 != 0 { 'w' } else { '-' });
    s.push(if mode & 0o100 != 0 { 'x' } else { '-' });
    s.push(if mode & 0o040 != 0 { 'r' } else { '-' });
    s.push(if mode & 0o020 != 0 { 'w' } else { '-' });
    s.push(if mode & 0o010 != 0 { 'x' } else { '-' });
    s.push(if mode & 0o004 != 0 { 'r' } else { '-' });
    s.push(if mode & 0o002 != 0 { 'w' } else { '-' });
    s.push(if mode & 0o001 != 0 { 'x' } else { '-' });
    s
}

fn process_size(size: u64) -> String {
    let units = ["B", "KB", "MB", "GB", "TB"];
    let mut s = size as f64;
    let mut u = 0usize;
    while s >= 1024.0 && u < 4 {
        s /= 1024.0;
        u += 1;
    }
    format!("{:.1} {}", s, units[u])
}

fn utf8_char_len(c: u8) -> usize {
    if c < 0x80 { 1 }
    else if (c >> 5) == 0b110 { 2 }
    else if (c >> 4) == 0b1110 { 3 }
    else if (c >> 3) == 0b11110 { 4 }
    else { 1 }
}

fn colorama_print(text: &str, col: u64) -> String {
    let bg_r = ((col >> 40) & 0xFF) as u8;
    let bg_g = ((col >> 32) & 0xFF) as u8;
    let bg_b = ((col >> 24) & 0xFF) as u8;
    let fg_r = ((col >> 16) & 0xFF) as u8;
    let fg_g = ((col >> 8)  & 0xFF) as u8;
    let fg_b = (col & 0xFF) as u8;
    format!("\x1b[38;2;{};{};{}m\x1b[48;2;{};{};{}m{}\x1b[0m",
            fg_r, fg_g, fg_b, bg_r, bg_g, bg_b, text)
}

fn colorama_fillup(text: &str, color: u64) -> String {
    let mut res = String::with_capacity(CARD_LEN);
    let mut pos = 0;
    let mut i = 0;
    let bytes = text.as_bytes();
    while pos < CARD_LEN && i < bytes.len() {
        let clen = utf8_char_len(bytes[i]);
        if pos + clen > CARD_LEN { break; }
        res.push_str(&text[i..i + clen]);
        pos += clen;
        i += clen;
    }
    while pos < CARD_LEN {
        res.push(' ');
        pos += 1;
    }
    colorama_print(&res, color)
}

fn coll(text: &str, col: u64) -> String {
    let r = ((col >> 16) & 0xFF) as u8;
    let g = ((col >> 8)  & 0xFF) as u8;
    let b = (col & 0xFF) as u8;
    format!("\x1b[38;2;{};{};{}m{}\x1b[0m", r, g, b, text)
}

fn chngg(text: &str, color: u64, more_space: bool) -> String {
    let max_len = if more_space { MAX_SIZE_ELEMENT * 3 } else { MAX_SIZE_ELEMENT };
    let mut res = String::with_capacity(max_len);
    let mut pos = 0;
    let mut i = 0;
    let bytes = text.as_bytes();
    while pos < max_len && i < bytes.len() {
        let clen = utf8_char_len(bytes[i]);
        if pos + clen > max_len { break; }
        res.push_str(&text[i..i + clen]);
        pos += clen;
        i += clen;
    }
    while pos < max_len {
        res.push(' ');
        pos += 1;
    }
    coll(&res, color)
}

fn get_type_ind_by_name(type_s: &str) -> u32 {
    unsafe {
        let arr = &raw const TYPES;
        for (i, t) in (*arr).iter().enumerate().skip(1) {
            if t.name == type_s {
                return i as u32;
            }
        }
    }
    ERRR
}

fn add_type_value(pp_index: usize, filesize: u64) {
    if pp_index >= NR_OF_SUPPORTED_LANG { return; }
    unsafe {
        let arr = &raw mut TYPES;
        (*arr)[pp_index].counted_size += filesize;
    }
}

fn get_type(fullpath: &str, filesize: u64) {
    if let Some(dot_pos) = fullpath.rfind('.') {
        let ext = &fullpath[dot_pos + 1..];
        if !ext.contains('/') {
            let idx = get_type_ind_by_name(ext);
            if idx != ERRR {
                add_type_value(idx as usize, filesize);
            }
        }
    }
}

fn folder_size(path: &Path) -> u64 {
    let mut total = 0u64;
    if let Ok(entries) = fs::read_dir(path) {
        for e in entries.filter_map(Result::ok) {
            let p = e.path();
            if let Ok(meta) = fs::symlink_metadata(&p) {
                if meta.is_dir() {
                    total += folder_size(&p);
                } else {
                    let sz = meta.len();
                    total += sz;
                    if let Some(s) = p.to_str() {
                        get_type(s, sz);
                    }
                }
            }
        }
    }
    total
}

fn get_dom_lang_index() -> usize {
    let mut local_sums = [0u64; SUPPORTED_GROUPS];
    unsafe {
        let arr = &raw const TYPES;
        for i in 1..NR_OF_SUPPORTED_LANG {
            let t = &(*arr)[i];
            let g = t.globalsumindex;
            if g < SUPPORTED_GROUPS {
                local_sums[g] += t.counted_size;
            }
        }
    }
    let mut max_idx = 0;
    let mut max_size = 0u64;
    for (i, &s) in local_sums.iter().enumerate() {
        if s > max_size {
            max_size = s;
            max_idx = i;
        }
    }
    unsafe {
        let arr = &raw mut TYPES;
        for t in &mut (*arr) {
            t.counted_size = 0;
        }
    }
    max_idx
}

fn get_dom_single_file_index(path: &Path) -> usize {
    let path_str = path.to_string_lossy();
    let extra = if let Some(pos) = path_str.rfind('.') {
        &path_str[pos + 1..]
    } else {
        ""
    };
    unsafe {
        let types = &raw const TYPES;
        for (i, t) in (*types).iter().enumerate() {
            if t.name == extra {
                return i;
            }
        }
    }
    0
}

fn main_lll() -> Result<(), Box<dyn std::error::Error>> {
    let mut entries: Vec<(u64, String)> = vec![];
    for entry in fs::read_dir(".")? {
        let entry = entry?;
        let path = entry.path();
        let name = entry.file_name().to_string_lossy().into_owned();
        if name == "." || name == ".." { continue; }
        let meta = fs::symlink_metadata(&path)?;
        let mode = meta.permissions().mode();
        let perms = mode_to_str(mode);
        let links = meta.nlink();
        let owner = users::get_user_by_uid(meta.uid())
            .map(|u| u.name().to_string_lossy().into_owned())
            .unwrap_or_else(|| format!("uid:{}", meta.uid()));
        let group = users::get_group_by_gid(meta.gid())
            .map(|g| g.name().to_string_lossy().into_owned())
            .unwrap_or_else(|| format!("gid:{}", meta.gid()));
        let size_str = if meta.is_dir() {
            process_size(folder_size(&path))
        } else {
            process_size(meta.len())
        };
        let mtime = meta.modified().unwrap_or(std::time::SystemTime::UNIX_EPOCH);
        let timestr = chrono::DateTime::<chrono::Local>::from(mtime)
            .format("%e. %b %H:%M")
            .to_string();
        let lang_idx;
        if meta.is_dir() {
            lang_idx = get_dom_lang_index();
        } else {
            lang_idx = get_dom_single_file_index(&path);
        }
        let is_dir = meta.is_dir();
        let name_color = if is_dir { 0x9999FF } else { 0xFF33FF };
        let line = if meta.is_symlink() {
            if let Ok(target) = fs::read_link(&path) {
                format!("{} {:2} {} {} {:8} {} {} -> {}",
                    perms, links, owner, group, size_str, timestr, name, target.display())
            } else {
                format!("{} {:2} {} {} {:8} {} {}", perms, links, owner, group, size_str, timestr, name)
            }
        } else {
            format!("{} {:2} {} {} {:8} {} {} {}",
                perms,
                links,
                chngg(&owner,     0x44FF44, false),
                chngg(&group,     0x88FF88, false),
                chngg(&size_str,  0xFF0000, false),
                chngg(&timestr,   0x888866, false),
                colorama_fillup(NAME_LIST[lang_idx], COLOR_LIST[lang_idx]),
                chngg(&name, name_color, true)
            )
        };
        let ts = mtime.duration_since(std::time::UNIX_EPOCH).unwrap_or_default().as_secs();
        entries.push((ts, line));
    }
    entries.sort_by(|a, b| b.0.cmp(&a.0));
    entries.reverse();
    for (_, line) in entries {
        println!("{}", line);
    }
    Ok(())
}

fn get_absolute_path(name: &str) {
    match env::current_dir() {
        Ok(mut path) => {
            path.push(name);
            if let Some(path_str) = path.to_str() {
                println!("{}", path_str);
                clipper(path_str);
            }
        }
        Err(e) => eprintln!("ERROR: getcwd: {}", e),
    }
}

fn clipper(content: &str) {
    let mut child = Command::new("xclip")
        .arg("-selection")
        .arg("clipboard")
        .stdin(Stdio::piped())
        .spawn()
        .expect("Fehler: xclip konnte nicht gestartet werden");
    if let Some(mut stdin) = child.stdin.take() {
        stdin.write_all(content.as_bytes()).unwrap();
    }
    let _ = child.wait(); // Warten, bis xclip fertig ist
}

fn main() {
    if env::args().len() == 1 {
        if let Err(e) = main_lll() {
            eprintln!("Error: {}", e);
        }
    }
    if env::args().len() == 2 {
        if let Some(name) = env::args().nth(1) {
            get_absolute_path(&name);
        } else {
            eprintln!("Bitte einen Namen als Parameter übergeben.");
        }
    }
}

